packages/extension-ui/src/components/MnemonicInput.tsx

```diff
@@ -0,0 +1,168 @@
+// Copyright 2019-2023 @polkadot/extension-ui authors & contributors
+// SPDX-License-Identifier: Apache-2.0
+
+import type { ThemeProps } from '../types';
+
+import React, { useCallback, useEffect, useState } from 'react';
+import styled from 'styled-components';
+
+import { validateSeed } from '@polkadot/extension-ui/messaging';
+import { objectSpread } from '@polkadot/util';
+import { KeypairType } from '@polkadot/util-crypto/types';
+
+import { Warning } from '../components';
+import useTranslation from '../hooks/useTranslation';
+import { AccountInfo } from '../Popup/ImportSeed';
+import MnemonicPill from './MnemonicPill';
+
+interface Props extends ThemeProps {
+  className?: string;
+  onAccountChange: (account: AccountInfo | null) => void;
+  type: KeypairType;
+  seed: string | null;
+  genesis: string;
+  path: string | null;
+  address: string;
+  setAddress: (address: string) => void;
+  error: string;
+  setError: (error: string) => void;
+  onChange: (seed: string) => void;
+}
+
+const MnemonicInput = ({
+  address,
+  className,
+  error,
+  genesis,
+  onAccountChange,
+  onChange,
+  path,
+  seed,
+  setAddress,
+  setError,
+  type
+}: Props) => {
+  const [mnemonicWords, setMnemonicWords] = useState<string[]>(Array(12).fill(''));
+  const { t } = useTranslation();
+
+  const _handlePaste = useCallback(
+    (event: React.ClipboardEvent) => {
+      event.preventDefault();
+      const MNEMONIC_LENGTH = 12;
+
+      setMnemonicWords([]);
+      const pastedWords = event.clipboardData.getData('text').trim().split(' ').slice(0, 12);
+      const defaultValue = '';
+      const fillArray = Array(MNEMONIC_LENGTH - pastedWords.length).fill(defaultValue);
+      const result = pastedWords.concat(fillArray);
```

Comment:
Is this for if someone pastes just, let's say 4 phrases instead of 12 you put those 4 + add undefined to satisfy the length for mapping?
Wouldn't it be easier to create and object with 12 numbers as keys? On change just modify each value similar as now via index number - and on paste update each world at next object value?  You don't have to keep array always the right length. It also gives the advantage that you could probably check in what field user pasted and set only words from that field until the end. For the case that I wrote first 8 and want to paste last two 🤔